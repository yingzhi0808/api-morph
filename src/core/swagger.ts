import { createRequire } from "node:module";
import path from "node:path";

/**
 * generateSwaggerUI 的选项配置
 *
 * @category Core
 */
export interface GenerateSwaggerUIOptions {
  /** Swagger JSON 文件的 URL 路径，默认为 '/openapi.json' */
  url?: string;
  /** HTML 页面标题，默认为 'Swagger UI' */
  title?: string;
  /** 自定义 CSS 样式 */
  customCss?: string;
  /** 自定义 JavaScript 代码 */
  customJs?: string;
  /** If set to true, it persists authorization data and it would not be lost on browser close/refresh。默认为 false */
  persistAuthorization?: boolean;
}

/**
 * 静态资源信息
 *
 * @category Core
 */
export interface SwaggerUIAssetInfo {
  /** swagger-ui-dist 包的绝对路径 */
  assetPath: string;
  /** 主要资源文件路径 */
  files: {
    indexCss: string;
    css: string;
    bundleJs: string;
    standalonePresetJs: string;
    favicon32: string;
    favicon16: string;
  };
}

/**
 * 获取 swagger-ui-dist 包的路径。
 * @returns swagger-ui-dist 包的绝对路径。
 */
function getSwaggerUIAssetPath() {
  const require = createRequire(import.meta.url);
  const packageJsonPath = require.resolve("swagger-ui-dist/package.json");
  return path.dirname(packageJsonPath);
}

/**
 * 获取 Swagger UI 静态资源信息。
 * @returns 包含资源路径和文件信息的对象。
 *
 * @category Core
 */
export function getSwaggerUIAssetInfo(): SwaggerUIAssetInfo {
  const assetPath = getSwaggerUIAssetPath();

  return {
    assetPath,
    files: {
      indexCss: path.join(assetPath, "index.css"),
      css: path.join(assetPath, "swagger-ui.css"),
      bundleJs: path.join(assetPath, "swagger-ui-bundle.js"),
      standalonePresetJs: path.join(assetPath, "swagger-ui-standalone-preset.js"),
      favicon32: path.join(assetPath, "favicon-32x32.png"),
      favicon16: path.join(assetPath, "favicon-16x16.png"),
    },
  };
}

/**
 * 生成 Swagger UI 的 HTML 字符串。
 * @param options wagger UI 配置选项。
 * @returns 完整的 Swagger UI HTML 字符串。
 *
 * @category Core
 */
export function generateSwaggerUI(options: GenerateSwaggerUIOptions = {}) {
  const {
    url = "/openapi.json",
    title = "Swagger UI",
    customCss = "",
    customJs = "",
    persistAuthorization = false,
  } = options;

  const html = `<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>${title}</title>
  <link rel="icon" type="image/png" href="./favicon-32x32.png" sizes="32x32">
  <link rel="icon" type="image/png" href="./favicon-16x16.png" sizes="16x16">
  <link rel="stylesheet" href="./swagger-ui.css">
  <link rel="stylesheet" href="./index.css">
  ${customCss ? `<style>\n${customCss}\n</style>` : ""}
</head>
<body>
  <div id="swagger-ui"></div>
  <script src="./swagger-ui-bundle.js"></script>
  <script src="./swagger-ui-standalone-preset.js"></script>
  <script>
    window.onload = () => {
      const ui = SwaggerUIBundle({
        url: "${url}",
        dom_id: "#swagger-ui",
        deepLinking: true,
        presets: [
          SwaggerUIBundle.presets.apis,
          SwaggerUIStandalonePreset
        ],
        plugins: [
          SwaggerUIBundle.plugins.DownloadUrl
        ],
        layout: "StandaloneLayout",
        persistAuthorization: ${persistAuthorization}
      });

      window.ui = ui;
    };
    ${customJs}
  </script>
</body>
</html>`;

  return html;
}
