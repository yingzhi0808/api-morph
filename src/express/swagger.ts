import type { Express } from "express";
import express from "express";
import {
  type GenerateSwaggerUIHTMLOptions,
  generateSwaggerUIHTML,
  getSwaggerUIAssetDir,
} from "@/core/swagger";

/**
 * 设置 SwaggerUI 的静态资源和路由
 * @param path SwaggerUI 路径
 * @param app Express 应用实例
 * @param options 选项
 *
 * @category Express
 */
export function setupSwaggerUI(
  path: string,
  app: Express,
  options: GenerateSwaggerUIHTMLOptions = {},
) {
  const assetDir = getSwaggerUIAssetDir();

  app.use(express.static(assetDir));

  app.get(path, (_req, res) => {
    const html = generateSwaggerUIHTML(options);
    res.setHeader("Content-Type", "text/html");
    res.send(html);
  });
}
