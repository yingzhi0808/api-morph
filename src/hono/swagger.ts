import { relative } from "node:path";
import { serveStatic } from "@hono/node-server/serve-static";
import type { Hono } from "hono";
import {
  type GenerateSwaggerUIHTMLOptions,
  generateSwaggerUIHTML,
  getSwaggerUIAssetDir,
} from "@/core/swagger";

/**
 * 设置 SwaggerUI 的静态资源和路由
 * @param path SwaggerUI 路径
 * @param app Hono 应用实例
 * @param options 选项
 *
 * @category Hono
 */
export function setupSwaggerUI(
  path: string,
  app: Hono,
  options: GenerateSwaggerUIHTMLOptions = {},
) {
  const assetDir = getSwaggerUIAssetDir();
  const relativePath = relative(process.cwd(), assetDir);
  app.use("/*", serveStatic({ root: relativePath }));

  app.get(path, (c) => {
    const html = generateSwaggerUIHTML(options);
    return c.html(html);
  });
}
