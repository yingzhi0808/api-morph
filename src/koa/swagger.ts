import type Koa from "koa";
import serve from "koa-static";
import {
  type GenerateSwaggerUIHTMLOptions,
  generateSwaggerUIHTML,
  getSwaggerUIAssetDir,
} from "@/core/swagger";

/**
 * 设置 SwaggerUI 的静态资源和路由
 * @param path SwaggerUI 路径
 * @param app Koa 应用实例
 * @param options 选项
 *
 * @category Koa
 */
export function setupSwaggerUI(path: string, app: Koa, options: GenerateSwaggerUIHTMLOptions = {}) {
  const assetDir = getSwaggerUIAssetDir();

  app.use(serve(assetDir));

  app.use(async (ctx, next) => {
    if (ctx.path === path && ctx.method === "GET") {
      const html = generateSwaggerUIHTML(options);
      ctx.set("Content-Type", "text/html");
      ctx.body = html;
    } else {
      await next();
    }
  });
}
