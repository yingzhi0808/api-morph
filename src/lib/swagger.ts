import { createRequire } from "node:module";
import path from "node:path";
import type { GenerateSwaggerUIOptions, SwaggerUIAssetInfo } from "@/types";

/**
 * 获取 swagger-ui-dist 包的路径。
 * @returns swagger-ui-dist 包的绝对路径。
 */
function getSwaggerUIAssetPath() {
  const require = createRequire(import.meta.url);
  const packageJsonPath = require.resolve("swagger-ui-dist/package.json");
  return path.dirname(packageJsonPath);
}

/**
 * 获取 Swagger UI 静态资源信息。
 * @returns 包含资源路径和文件信息的对象。
 */
export function getSwaggerUIAssetInfo(): SwaggerUIAssetInfo {
  const assetPath = getSwaggerUIAssetPath();

  return {
    assetPath,
    files: {
      indexCss: path.join(assetPath, "index.css"),
      css: path.join(assetPath, "swagger-ui.css"),
      bundleJs: path.join(assetPath, "swagger-ui-bundle.js"),
      standalonePresetJs: path.join(assetPath, "swagger-ui-standalone-preset.js"),
      favicon32: path.join(assetPath, "favicon-32x32.png"),
      favicon16: path.join(assetPath, "favicon-16x16.png"),
    },
  };
}

/**
 * 生成 Swagger UI 的 HTML 字符串。
 * @param options wagger UI 配置选项。
 * @returns 完整的 Swagger UI HTML 字符串。
 */
export function generateSwaggerUI(options: GenerateSwaggerUIOptions = {}) {
  const {
    url = "/openapi.json",
    title = "Swagger UI",
    customCss = "",
    customJs = "",
    persistAuthorization = false,
  } = options;

  const html = `<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>${title}</title>
  <link rel="icon" type="image/png" href="./favicon-32x32.png" sizes="32x32">
  <link rel="icon" type="image/png" href="./favicon-16x16.png" sizes="16x16">
  <link rel="stylesheet" href="./swagger-ui.css">
  <link rel="stylesheet" href="./index.css">
  ${customCss ? `<style>\n${customCss}\n</style>` : ""}
</head>
<body>
  <div id="swagger-ui"></div>
  <script src="./swagger-ui-bundle.js"></script>
  <script src="./swagger-ui-standalone-preset.js"></script>
  <script>
    window.onload = () => {
      const ui = SwaggerUIBundle({
        url: "${url}",
        dom_id: "#swagger-ui",
        deepLinking: true,
        presets: [
          SwaggerUIBundle.presets.apis,
          SwaggerUIStandalonePreset
        ],
        plugins: [
          SwaggerUIBundle.plugins.DownloadUrl
        ],
        layout: "StandaloneLayout",
        persistAuthorization: ${persistAuthorization}
      });

      window.ui = ui;
    };
    ${customJs}
  </script>
</body>
</html>`;

  return html;
}
