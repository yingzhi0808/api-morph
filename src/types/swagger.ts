/**
 * generateSwaggerUI 的选项配置
 */
export interface GenerateSwaggerUIOptions {
  /** Swagger JSON 文件的 URL 路径，默认为 '/openapi.json' */
  url?: string;
  /** HTML 页面标题，默认为 'Swagger UI' */
  title?: string;
  /** 自定义 CSS 样式 */
  customCss?: string;
  /** 自定义 JavaScript 代码 */
  customJs?: string;
  /** If set to true, it persists authorization data and it would not be lost on browser close/refresh。默认为 false */
  persistAuthorization?: boolean;
}

/**
 * 静态资源信息
 */
export interface SwaggerUIAssetInfo {
  /** swagger-ui-dist 包的绝对路径 */
  assetPath: string;
  /** 主要资源文件路径 */
  files: {
    indexCss: string;
    css: string;
    bundleJs: string;
    standalonePresetJs: string;
    favicon32: string;
    favicon16: string;
  };
}
