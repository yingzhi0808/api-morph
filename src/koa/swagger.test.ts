import Koa from "koa";
import request from "supertest";
import { describe, expect, it } from "vitest";
import type { GenerateSwaggerUIHTMLOptions } from "@/core/swagger";
import { setupSwaggerUI } from "./swagger";

describe("setupSwaggerUI", () => {
  it("应该使用默认选项设置Swagger", async () => {
    const app = new Koa();
    const path = "/api-docs";

    setupSwaggerUI(path, app);

    const response = await request(app.callback())
      .get(path)
      .expect(200)
      .expect("Content-Type", /text\/html/);

    expect(response.text).toMatchSnapshot();
  });

  it("应该使用自定义选项设置Swagger", async () => {
    const app = new Koa();
    const path = "/docs";
    const options: GenerateSwaggerUIHTMLOptions = {
      url: "/api/openapi.json",
      title: "My API Documentation",
      customCss: "body { color: red; }",
      customJs: "console.log('test');",
      persistAuthorization: true,
    };

    setupSwaggerUI(path, app, options);

    const response = await request(app.callback())
      .get(path)
      .expect(200)
      .expect("Content-Type", /text\/html/);

    expect(response.text).toMatchSnapshot();
  });

  it("应该处理空对象选项", async () => {
    const app = new Koa();
    const path = "/docs";
    const options = {};

    setupSwaggerUI(path, app, options);

    const response = await request(app.callback())
      .get(path)
      .expect(200)
      .expect("Content-Type", /text\/html/);

    expect(response.text).toMatchSnapshot();
  });

  it("应该在不同路径下正确工作", async () => {
    const app = new Koa();

    // 设置多个不同的 Swagger UI 路径
    setupSwaggerUI("/api-docs", app, { title: "API Docs" });
    setupSwaggerUI("/swagger", app, { title: "Swagger" });

    // 测试第一个路径
    const response1 = await request(app.callback()).get("/api-docs").expect(200);
    expect(response1.text).toContain("API Docs");

    // 测试第二个路径
    const response2 = await request(app.callback()).get("/swagger").expect(200);
    expect(response2.text).toContain("Swagger");
  });
});
