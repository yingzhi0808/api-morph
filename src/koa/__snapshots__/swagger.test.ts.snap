// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`setupSwaggerUI > 应该使用自定义选项设置Swagger 1`] = `
"<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>My API Documentation</title>
  <link rel="icon" type="image/png" href="./favicon-32x32.png" sizes="32x32">
  <link rel="icon" type="image/png" href="./favicon-16x16.png" sizes="16x16">
  <link rel="stylesheet" href="./swagger-ui.css">
  <link rel="stylesheet" href="./index.css">
  <style>
body { color: red; }
</style>
</head>
<body>
  <div id="swagger-ui"></div>
  <script src="./swagger-ui-bundle.js"></script>
  <script src="./swagger-ui-standalone-preset.js"></script>
  <script>
    window.onload = () => {
      const ui = SwaggerUIBundle({
        url: "/api/openapi.json",
        dom_id: "#swagger-ui",
        deepLinking: true,
        presets: [
          SwaggerUIBundle.presets.apis,
          SwaggerUIStandalonePreset
        ],
        plugins: [
          SwaggerUIBundle.plugins.DownloadUrl
        ],
        layout: "StandaloneLayout",
        persistAuthorization: true
      });

      window.ui = ui;
    };
    console.log('test');
  </script>
</body>
</html>"
`;

exports[`setupSwaggerUI > 应该使用默认选项设置Swagger 1`] = `
"<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>Swagger UI</title>
  <link rel="icon" type="image/png" href="./favicon-32x32.png" sizes="32x32">
  <link rel="icon" type="image/png" href="./favicon-16x16.png" sizes="16x16">
  <link rel="stylesheet" href="./swagger-ui.css">
  <link rel="stylesheet" href="./index.css">
  
</head>
<body>
  <div id="swagger-ui"></div>
  <script src="./swagger-ui-bundle.js"></script>
  <script src="./swagger-ui-standalone-preset.js"></script>
  <script>
    window.onload = () => {
      const ui = SwaggerUIBundle({
        url: "/openapi.json",
        dom_id: "#swagger-ui",
        deepLinking: true,
        presets: [
          SwaggerUIBundle.presets.apis,
          SwaggerUIStandalonePreset
        ],
        plugins: [
          SwaggerUIBundle.plugins.DownloadUrl
        ],
        layout: "StandaloneLayout",
        persistAuthorization: false
      });

      window.ui = ui;
    };
    
  </script>
</body>
</html>"
`;

exports[`setupSwaggerUI > 应该处理空对象选项 1`] = `
"<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>Swagger UI</title>
  <link rel="icon" type="image/png" href="./favicon-32x32.png" sizes="32x32">
  <link rel="icon" type="image/png" href="./favicon-16x16.png" sizes="16x16">
  <link rel="stylesheet" href="./swagger-ui.css">
  <link rel="stylesheet" href="./index.css">
  
</head>
<body>
  <div id="swagger-ui"></div>
  <script src="./swagger-ui-bundle.js"></script>
  <script src="./swagger-ui-standalone-preset.js"></script>
  <script>
    window.onload = () => {
      const ui = SwaggerUIBundle({
        url: "/openapi.json",
        dom_id: "#swagger-ui",
        deepLinking: true,
        presets: [
          SwaggerUIBundle.presets.apis,
          SwaggerUIStandalonePreset
        ],
        plugins: [
          SwaggerUIBundle.plugins.DownloadUrl
        ],
        layout: "StandaloneLayout",
        persistAuthorization: false
      });

      window.ui = ui;
    };
    
  </script>
</body>
</html>"
`;
